assert/assertion_compare.go

```diff
@@ -31,8 +51,8 @@ func compare(obj1, obj2 interface{}, kind reflect.Kind) (CompareType, bool) {
 		}
 	case reflect.Int8:
 		{
-			int8obj1 := obj1.(int8)
-			int8obj2 := obj2.(int8)
+			int8obj1 := obj1Value.Convert(int8Type).Interface().(int8)
```

Comment:
Would these calls to `Convert()` not actually modify the type being tested and as such, have side-effects?

Reply:
Consider this snippet: 

```go
type myInt int
assert.Less(t, int(1), myInt(2))
```

In my mind, this should work, and return `true`. Right now it panics. Is your point above that this snippet should not work, because they are different types?

Reply:
I see what you mean. Makes sense to me then. Just slightly worried that the reflection would have a non-negligible effect on performance. 

Would you mind running a benchmark (using `testing.B`) to compare the before and after on the working (i.e. non-custom type) cases?

Reply:
I'm not much experienced with benchmarks, so LMK if I'm doing something wrong here. But here's an attempt: 

My PR, with TestCompare renamed to BenchmarkCompare, left as-is prior to this PR, except wrapping the body in `for i := 0; i < t.N; i++ { <test here> }`:

```
Running tool: /Users/torkel/go/go1.14.1/bin/go test -benchmem -run=^$ github.com/stretchr/testify/assert -bench ^(BenchmarkCompare)$

goos: darwin
goarch: amd64
pkg: github.com/stretchr/testify/assert
BenchmarkCompare-8   	  437868	      2691 ns/op	     312 B/op	      66 allocs/op
PASS
ok  	github.com/stretchr/testify/assert	1.317
```

Master: 
```
Running tool: /Users/torkel/go/go1.14.1/bin/go test -benchmem -run=^$ github.com/stretchr/testify/assert -bench ^(BenchmarkCompare)$

goos: darwin
goarch: amd64
pkg: github.com/stretchr/testify/assert
BenchmarkCompare-8   	 4445959	       282 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/stretchr/testify/assert	1.880s
```

That's a quite unfortunate difference in performance... Might be possible to optimize this a bit? I'll take a look